File: backend-node/services/imageGenerator.js
import axios from 'axios';
import fs from 'fs';
import path from 'path';
import sharp from 'sharp';

export async function generateImages(jobId, script, duration) {
  const imagePaths = [];
  const scenes = script.scenes || [{ visual: script.hook || 'inspiring scene' }];
  
  for (let idx = 0; idx < scenes.length; idx++) {
    const scene = scenes[idx];
    const visualPrompt = scene.visual || 'inspiring background';
    
    // Enhanced prompt for better graphics
    const enhancedPrompt = `${visualPrompt}, vibrant colors, cinematic lighting, high quality, dramatic, 4k, professional photography`;
    const prompt = encodeURIComponent(enhancedPrompt);
    
    // Pollinations.ai with enhanced parameters
    const url = `https://image.pollinations.ai/prompt/${prompt}?width=1080&height=1920&nologo=true&enhance=true`;
    
    try {
      const response = await axios.get(url, {
        responseType: 'arraybuffer',
        timeout: 30000
      });
      
      if (response.status === 200) {
        // Resize and enhance to 1080x1920 (9:16 aspect ratio)
        const imageBuffer = await sharp(response.data)
          .resize(1080, 1920, { fit: 'cover' })
          .modulate({
            brightness: 1.1,
            saturation: 1.2
          })
          .sharpen()
          .jpeg({ quality: 95 })
          .toBuffer();
        
        const imagePath = path.join('temp', `${jobId}_img_${idx}.jpg`);
        fs.writeFileSync(imagePath, imageBuffer);
        imagePaths.push(imagePath);
      } else {
        imagePaths.push(await createFallbackImage(jobId, idx));
      }
      
    } catch (error) {
      console.error(`Image generation failed for scene ${idx}:`, error.message);
      imagePaths.push(await createFallbackImage(jobId, idx));
    }
    
    // Rate limiting
    await new Promise(resolve => setTimeout(resolve, 1000));
  }
  
  return imagePaths;
}

async function createFallbackImage(jobId, idx) {
  const imagePath = path.join('temp', `${jobId}_img_${idx}.jpg`);
  
  // Create vibrant gradient background
  const colors = [
    { r: 138, g: 43, b: 226 },  // Purple
    { r: 255, g: 20, b: 147 },  // Pink
    { r: 255, g: 140, b: 0 },   // Orange
    { r: 0, g: 191, b: 255 }    // Blue
  ];
  
  const color = colors[idx % colors.length];
  
  // Create gradient using SVG
  const svg = `
    <svg width="1080" height="1920">
      <defs>
        <linearGradient id="grad" x1="0%" y1="0%" x2="100%" y2="100%">
          <stop offset="0%" style="stop-color:rgb(${color.r},${color.g},${color.b});stop-opacity:1" />
          <stop offset="100%" style="stop-color:rgb(${Math.floor(color.r*0.5)},${Math.floor(color.g*0.5)},${Math.floor(color.b*0.5)});stop-opacity:1" />
        </linearGradient>
      </defs>
      <rect width="1080" height="1920" fill="url(#grad)" />
    </svg>
  `;
  
  await sharp(Buffer.from(svg))
    .jpeg()
    .toFile(imagePath);
  
  return imagePath;
}

File: backend/services/image_generator.py
import requests
from pathlib import Path
from PIL import Image
from io import BytesIO
import time


async def generate_images(job_id: str, script: dict, duration: int) -> list:
    """Generate images for each scene using Pollinations.ai"""
    
    image_paths = []
    scenes = script.get("scenes", [])
    
    if not scenes:
        # Fallback: create one image
        scenes = [{"visual": script.get("hook", "inspiring scene")}]
    
    for idx, scene in enumerate(scenes):
        visual_prompt = scene.get("visual", "inspiring background")
        
        # Clean prompt for URL
        prompt = visual_prompt.replace(" ", "%20")
        
        # Pollinations.ai - free, no API key
        url = f"https://image.pollinations.ai/prompt/{prompt}?width=1080&height=1920&nologo=true"
        
        try:
            response = requests.get(url, timeout=30)
            
            if response.status_code == 200:
                img = Image.open(BytesIO(response.content))
                
                # Ensure 9:16 aspect ratio (1080x1920)
                img = img.resize((1080, 1920), Image.Resampling.LANCZOS)
                
                image_path = Path("temp") / f"{job_id}_img_{idx}.jpg"
                img.save(image_path, "JPEG", quality=95)
                image_paths.append(image_path)
            else:
                # Fallback: solid color image
                image_paths.append(create_fallback_image(job_id, idx))
                
        except Exception as e:
            print(f"Image generation failed for scene {idx}: {e}")
            image_paths.append(create_fallback_image(job_id, idx))
        
        # Rate limiting
        time.sleep(1)
    
    return image_paths


def create_fallback_image(job_id: str, idx: int) -> Path:
    """Create a solid color fallback image"""
    from PIL import Image, ImageDraw, ImageFont
    
    img = Image.new('RGB', (1080, 1920), color=(30, 30, 50))
    draw = ImageDraw.Draw(img)
    
    # Add text
    try:
        font = ImageFont.truetype("arial.ttf", 60)
    except:
        font = ImageFont.load_default()
    
    text = f"Scene {idx + 1}"
    bbox = draw.textbbox((0, 0), text, font=font)
    text_width = bbox[2] - bbox[0]
    text_height = bbox[3] - bbox[1]
    
    position = ((1080 - text_width) // 2, (1920 - text_height) // 2)
    draw.text(position, text, fill=(255, 255, 255), font=font)
    
    image_path = Path("temp") / f"{job_id}_img_{idx}.jpg"
    img.save(image_path, "JPEG")
    
    return image_path
